File: agentic-patterns/js/src/sequential-processing.ts
import { z } from 'genkit';
import { googleAI } from '@genkit-ai/google-genai';
import { ai } from './genkit.js';

export const storyWriterFlow = ai.defineFlow(
  {
    name: 'storyWriterFlow',
    inputSchema: z.object({ topic: z.string() }),
    outputSchema: z.string(),
  },
  async ({ topic }) => {
    // Step 1: Generate a creative story idea
    const ideaResponse = await ai.generate({
      prompt: `Generate a unique story idea about a ${topic}.`,
      output: {
        schema: z.object({
          idea: z.string().describe('A short, compelling story concept'),
        }),
      },
    });

    const storyIdea = ideaResponse.output?.idea;
    if (!storyIdea) {
      throw new Error('Failed to generate a story idea.');
    }

    // Step 2: Use the idea to write the beginning of the story
    const storyResponse = await ai.generate({
      prompt: `Write the opening paragraph for a story based on this idea: ${storyIdea}`,
    });

    return storyResponse.text;
  }
);

export const imageGeneratorFlow = ai.defineFlow(
  {
    name: 'imageGeneratorFlow',
    inputSchema: z.object({ concept: z.string() }),
    outputSchema: z.string(),
  },
  async ({ concept }) => {
    // Step 1: Use a text model to generate a rich image prompt
    const promptResponse = await ai.generate({
      model: googleAI.model('gemini-2.5-flash'),
      prompt: `Create a detailed, artistic prompt for an image generation model. The concept is: "${concept}".`,
    });

    const imagePrompt = promptResponse.text;

    // Step 2: Use the generated prompt to create an image
    const imageResponse = await ai.generate({
      model: googleAI.model('imagen-3.0-generate-002'),
      prompt: imagePrompt,
      output: { format: 'media' },
    });

    const imageUrl = imageResponse.media?.url;
    if (!imageUrl) {
      throw new Error('Failed to generate an image.');
    }
    return imageUrl;
  }
);

File: agentic-patterns/dart/lib/sequential_processing.dart
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import 'package:genkit/genkit.dart';
import 'package:genkit_google_genai/genkit_google_genai.dart';
import 'package:schemantic/schemantic.dart';

part 'sequential_processing.g.dart';

@Schema()
abstract class $StoryInput {
  @Field(defaultValue: 'dinosaurs')
  String get topic;
}

@Schema()
abstract class $StoryIdea {
  /// A short, compelling story concept
  String get idea;
}

@Schema()
abstract class $ImageGeneratorInput {
  @Field(defaultValue: 'a futuristic city')
  String get concept;
}

void defineSequentialProcessingFlows(Genkit ai) {
  ai.defineFlow(
    name: 'storyWriterFlow',
    inputSchema: StoryInput.$schema,
    outputSchema: .string(),
    fn: (input, _) async {
      // Step 1: Generate a creative story idea
      final ideaResponse = await ai.generate(
        prompt: 'Generate a unique story idea about a ${input.topic}.',
        outputSchema: StoryIdea.$schema,
      );

      final storyIdea = ideaResponse.output?.idea;
      if (storyIdea == null) {
        throw Exception('Failed to generate a story idea.');
      }

      // Step 2: Use the idea to write the beginning of the story
      final storyResponse = await ai.generate(
        prompt:
            'Write the opening paragraph for a story based on this idea: $storyIdea',
      );

      return storyResponse.text;
    },
  );

  ai.defineFlow(
    name: 'imageGeneratorFlow',
    inputSchema: ImageGeneratorInput.$schema,
    outputSchema: .string(),
    fn: (input, _) async {
      // Step 1: Use a text model to generate a rich image prompt
      final promptResponse = await ai.generate(
        prompt:
            'Create a detailed, artistic prompt for an image generation model. The concept is: "${input.concept}".',
      );

      final imagePrompt = promptResponse.text;

      // Step 2: Use the generated prompt to create an image
      final imageResponse = await ai.generate(
        model: googleAI.gemini('gemini-2.5-flash-image'),
        prompt: imagePrompt,
        config: {
          'responseModalities': ['image'],
        },
      );

      final imageUrl = imageResponse.media?.url;
      if (imageUrl == null) {
        throw Exception('Failed to generate an image.');
      }
      return imageUrl;
    },
  );
}
